### PR TITLE
82 add move function

### LINKED ISSUE #82: Add move function
Using the legal moves make a move function.

### PR DESCRIPTION
**Pull request by abrahimt**
Added the move piece function

### CODE REVIEW

File: cli/src/main.rs [hunk 1]
```diff
@@ -53,15 +53,35 @@ fn main() -> Result<(), Error> {
...
+            Arg::new("move")
+                .long("move")
+                .short('m')
+                .value_name("move")
+                .help("Make a move from starting square to final square (ex: 'E2:E4'"),
+        )
         .get_matches();
-
+    // Happen every time //
     let fen = matches
```
Comment (chanson02) [inline cli/src/main.rs]: nit: let's add some whitespace here.

File: cli/src/main.rs [hunk 2]
```diff
@@ -53,15 +53,35 @@ fn main() -> Result<(), Error> {
                 .help("Determines who is winning. Positive number indicates a white advantage.")
                 .action(clap::ArgAction::SetTrue),
         )
+        .arg(
+            Arg::new("move")
+                .long("move")
+                .short('m')
+                .value_name("move")
+                .help("Make a move from starting square to final square (ex: 'E2:E4'"),
```
Comment (chanson02) [inline cli/src/main.rs]: Does this help message make sense? Maybe just change it to `moves a piece`

File: cli/src/main.rs [hunk 3]
```diff
@@ -53,15 +53,35 @@ fn main() -> Result<(), Error> {
...
+            let from = position::string_to_square(start_square)
+                .map_err(|e| Error::ArgumentError(e.to_string()))?;
+            if let Some(end_square) = move_coords.next() {
+                let to = position::string_to_square(end_square)
+                    .map_err(|e| Error::ArgumentError(e.to_string()))?;
+                gs.move_piece_legally(from, to)
+                    .map_err(|e| Error::ArgumentError(e.to_string()))?;
+            }
+        }
+    }
```
Comment (chanson02) [inline cli/src/main.rs]: There has to be a cleaner way to write this.

File: engine/src/move_generation.rs [hunk 4]
```diff
@@ -676,4 +706,13 @@ mod tests {
...
     }
+
+    #[test]
+    fn test_move_piece() {
+        let mut gs = GameState::new();
+
+        gs.move_piece(12, 28);
+        assert_eq!(gs.board.piece_at_position(12), None);
+        assert_eq!(gs.board.piece_at_position(28), Some('P'));
+    }
```
Comment (chanson02) [inline engine/src/move_generation.rs]: I feel like we should also have tests for legal moving.

File: engine/src/move_generation.rs [hunk 5]
```diff
@@ -2,6 +2,36 @@ use super::GameState;
...
+            *from_piece_bitboard &= !(1 << from);
+            *from_piece_bitboard |= 1 << to;
+        }
+    }
+
+    pub fn move_piece_legally(&mut self, from: u8, to: u8) -> Result<(), String> {
+        // "E2:E4"
+        //Get moves from E2
+        let possible_moves = self.possible_moves(from);
+        //Compare E4 to get moves
```
Comment (chanson02) [inline engine/src/move_generation.rs]: Remove comments

File: engine/src/move_generation.rs [hunk 6]
```diff
@@ -2,6 +2,36 @@ use super::GameState;
...
+        if 1 << to & possible_moves == 0 {
+            return Err(("Not a legal move").to_string());
```
Comment (chanson02) [inline engine/src/move_generation.rs]: Maybe we could make this more specific. Like `E1 -> E2 move illegal`

**Review by chanson02 — CHANGES_REQUESTED** [review]
I can redo the nits myself in order to get them just the way I like them lol.

File: cli/src/main.rs (hunk 1 above)
Comment (abrahimt) [reply cli/src/main.rs]: Fixed

File: cli/src/main.rs (hunk 2 above)
Comment (abrahimt) [reply cli/src/main.rs]: Fixed

File: engine/src/move_generation.rs (hunk 4 above)
Comment (abrahimt) [reply engine/src/move_generation.rs]: Added

File: engine/src/move_generation.rs (hunk 5 above)
Comment (abrahimt) [reply engine/src/move_generation.rs]: Fixed

File: engine/src/move_generation.rs (hunk 6 above)
Comment (abrahimt) [reply engine/src/move_generation.rs]: Okay I'll work on that

Comment (abrahimt) [reply engine/src/move_generation.rs]: Done

File: cli/src/main.rs (hunk 3 above)
Comment (abrahimt) [reply cli/src/main.rs]: Bet

**Review by chanson02 — APPROVED** [review]

**Pull request merged**